sdk/src/runtime/lib/router.ts

```diff
@@ -167,27 +171,25 @@ export function defineRoutes<TContext = Record<string, any>>(
       }
 
       let { params, handler } = match;
+      routeContext.params = params;
 
-      let handlers = Array.isArray(handler) ? handler : [handler];
+      const handlers = Array.isArray(handler) ? handler : [handler];
       for (const h of handlers) {
         if (isRouteComponent(h)) {
-          const actionResult = await rw.handleAction(ctx);
+          const actionResult = await rw.handleAction(routeContext);
           const serializedEnv = serializeEnv(env);
           const props = { params, env: serializedEnv, ctx };
           return await rw.renderPage({
-            Page: handler as React.FC<Record<string, any>>,
+            Page: h as React.FC<Record<string, any>>,
             props,
             actionResult,
             Layout: rw.Layout,
           });
         } else {
-          return await (h({
-            request,
-            params,
-            ctx,
-            env,
-            rw,
-          }) as Promise<Response>);
+          const r = await (h(routeContext) as Promise<Response>);
+          if (r instanceof Response) {
+            return r;
+          }
```

Comment:
@peterp I needed to change this, otherwise middleware responses would short-circuit the chain, e.g. `index([() => { /* we would stop here */ }, Home])`

Reply:
🆒 🆒 🆒 

This probably makes more sense anyway... Otherwise we try and send things that are not valid over the wire.